### PR TITLE
[#2105] Hash session ID in the database

### PR DESCRIPTION
**Pull request by drupal-spider**
This pull request makes the following changes:

- Add a new private function to hash string.
- Hash session ID when set, get and delete session from the database.

### CODE REVIEW

File: src/SimpleSAML/Store/SQLStore.php [hunk 1]
```diff
@@ -436,4 +448,25 @@ public function delete(string $type, string $key): void
...
+
+
+    /**
+     * Calculates a base-64 encoded, URL-safe sha-256 hash.
+     *
+     * @param string $data
+     * @return string The hashed data.
+     */
+    public static function hashBase64(string $data) {
+        $hash = base64_encode(hash('sha256', $data, TRUE));
```
Comment (tvdijen) [inline src/SimpleSAML/Store/SQLStore.php:460]: May I suggest using the following instead?

```
$secretSalt = (new Utils\Config())->getSecretSalt();
return hash_hmac('sha256', $data, $secretSalt);
```

hash_hmac() will always produce a url-safe string.

Comment (tvdijen) [reply src/SimpleSAML/Store/SQLStore.php:460]: I also think this can be solved with a non-static private method, which is easier to unit-test.

Comment (monkeyiq) [reply src/SimpleSAML/Store/SQLStore.php:460]: Using the salt will also provide more privacy

Comment (drupal-spider) [reply src/SimpleSAML/Store/SQLStore.php:460]: Yes, that is a good idea.

**Comment by tvdijen** [comment]
Please fix the coding style and we're a go

**Comment by drupal-spider** [comment]
> Please fix the coding style and we're a go

Style has been fixed, please review.
Thanks.

**Comment by drupal-spider** [comment]
Rebase simplesamlphp:master

**Comment by tvdijen** [comment]
Merged and backported to the 2.3.x branch. Release is yet to be determined.

**Pull request merged**